### PR TITLE
Fix memory issue in mxfp8 model init

### PR DESCRIPTION
**Pull request by WanZzzzzz**
# What does this PR do ?
In current fp8_recipe=mxfp8, fp8_param_gather=true,  reuse_grad_buf_for_mxfp8_param_ag=true workflow, there will two buffer allocation for bf16 weights (laynorm, etc): one is for module parameter, another one is dist opt buffer allocation. There are also unnecessary copies for bf16 weights.
This PR addresses the above issue for the bf16 weights cases in a mxfp8 model. 
The change passed the test using test_mxfp8.py.

:warning: For major changes (either in lines of code or in its impact), please make sure to first share a design doc with the team. If you're unsure what's the best way to do so, contact the @mcore-oncall.

## Contribution process

```mermaid
flowchart LR
    A[Pre-checks] --> B[PR Tests]
    subgraph Code Review/Approval
        C1[Expert Review] --> C2[Final Review]
    end
    B --> C1
    C2 --> D[Merge]
```

### Pre-checks

- [ ] I want this PR in a versioned release and have added the appropriate Milestone (e.g., `Core 0.8`)
- [ ] I have added relevant unit tests
- [ ] I have added relevant functional tests
- [x] I have added proper typing to my code [Typing guidelines](https://docs.python.org/3/library/typing.html)
- [ ] I have added relevant documentation
- [x] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

### Code review

The following process is enforced via the CODEOWNERS file for changes into `megatron/core`. For changes outside of `megatron/core`, it is up to the PR author whether or not to tag the Final Reviewer team.

<details>
<summary>For MRs into `main` branch</summary>

Feel free to message or comment the @mcore-oncall to help accelerate your merge into main. The less complex your PR is, the faster it will be approved and merged!

#### (Step 2): Collect the expert reviewers reviews

1. Attach the `Expert Review` label when your PR is ready for review.
2. GitHub auto-assigns expert reviewers based on your changes. They will get notified and pick up your PR soon.

:warning: Only proceed to the next step once all reviewers have approved, merge-conflict are resolved and the CI is passing.  
Final Review might get declined if these requirements are not fulfilled.

#### (Step 3): Final Review

1. Add `Final Review` label
2. GitHub auto-assigns final reviewers based on your changes. They will get notified and pick up your PR soon.

#### (Optional Step 4): Cherry-pick into release branch

If this PR also needs to be merged into `core_r*` release branches, after this PR has been merged, select `Cherry-pick` to open a new PR into the release branch.

</details>

<details>
<summary>For MRs into `dev` branch</summary>
The proposed review process for `dev` branch is under active discussion.

MRs are mergable after one approval by either `eharper@nvidia.com` or `zijiey@nvidia.com`.
</details>

### Merging your PR

Any member of [core-adlr](https://github.com/orgs/teams/NVIDIA/core-adlr) and [`core-nemo`](https://github.com/orgs/teams/NVIDIA/core-nemo) will be able to merge your PR.

### CODE REVIEW

**Comment by Phlip79** [comment]
Please add further comments in the code explaining why the code breaks if `is_bf16_weight_group` is True.

**Comment by WanZzzzzz** [comment]
> Please add further comments in the code explaining why the code breaks if `is_bf16_weight_group` is True.

Thanks. Done.

**Comment by WanZzzzzz** [comment]
@ko3n1g @chtruong814 
I made changes according to the comments by @Phlip79 . What else needs to be done?

**Comment by ko3n1g** [comment]
/ok to test [bff91de](https://github.com/NVIDIA/Megatron-LM/pull/3461/commits/bff91de466ed52a93e703b25e5c75e0517225f08)

File: megatron/core/distributed/param_and_grad_buffer.py [hunk 1]
```diff
@@ -820,8 +828,21 @@ def _does_param_require_new_bucket(param):
         cur_bucket_id = 0
         for param in params[::-1]:
             param_start_index, param_end_index, bucket_id = self.param_index_map[param]
-            # For MXFP8 param: we only need to map weight gradients to the buffer.
-            if not self.ddp_config.reuse_grad_buf_for_mxfp8_param_ag:
+            # For MXFP8 param:
+            # we only need to map bf16 weights (layernorm, embedding, etc) to the buffer.
+            if self.ddp_config.reuse_grad_buf_for_mxfp8_param_ag:
+                if not is_mxfp8tensor(param) and not is_float8tensor(param):
+                    if self.param_data is not None:
```
Comment (deepakn94) [inline megatron/core/distributed/param_and_grad_buffer.py]: Should we make this a small helper function? Seems identical to code in the else block.

File: megatron/core/distributed/distributed_data_parallel.py [hunk 2]
```diff
@@ -488,9 +488,15 @@ def start_param_sync(self, *unused, force_sync: bool = False, force_dispatch: bo
                 # The paramaters are cast from bf16 to MXFP8 during copy.
                 # In the case of "overlap_param_gather=True", the param copy is done
                 # in "finish_param_sync" stage after zeroing the shared gardient buffers.
+                is_bf16_weight_group = False
                 if self.ddp_config.reuse_grad_buf_for_mxfp8_param_ag:
                     for bucket in bucket_group.buckets:
                         for param in bucket.params:
+                            # Skip copying since bf16 weights in the mxfp8 model
+                            # are already mapped to param.data.
+                            if not is_mxfp8tensor(param) and not is_float8tensor(param):
```
Comment (kunlunl) [inline megatron/core/distributed/distributed_data_parallel.py]: I think `not is_mxfp8tensor(param)` is redundant, `is_float8tensor` checks the base class of all fp8 tensors, so if `is_float8tensor` is False, then `is_mxfp8tensor` must also be False?

File: megatron/core/distributed/distributed_data_parallel.py [hunk 3]
```diff
@@ -488,9 +488,15 @@ def start_param_sync(self, *unused, force_sync: bool = False, force_dispatch: bo
                 # The paramaters are cast from bf16 to MXFP8 during copy.
                 # In the case of "overlap_param_gather=True", the param copy is done
                 # in "finish_param_sync" stage after zeroing the shared gardient buffers.
+                is_bf16_weight_group = False
```
Comment (kunlunl) [inline megatron/core/distributed/distributed_data_parallel.py]: Should this line be placed inside the `for bucket in bucket_group.buckets` loop? A bucket group may have bf16 bucket and fp8 bucket at the same time, so it should be reset for each bucket.

File: megatron/core/distributed/distributed_data_parallel.py [hunk 4]
```diff
@@ -7,7 +7,7 @@
 import torch
 
 from ..config_logger import has_config_logger_enabled, log_config_to_disk
-from ..fp8_utils import is_float8tensor, post_all_gather_processing
+from ..fp8_utils import is_float8tensor, is_mxfp8tensor, post_all_gather_processing
```
Comment (kunlunl) [inline megatron/core/distributed/distributed_data_parallel.py]: Redundant if `is_mxfp8tensor` is not used, see below.

File: megatron/core/distributed/distributed_data_parallel.py [hunk 5]
```diff
@@ -500,6 +506,8 @@ def start_param_sync(self, *unused, force_sync: bool = False, force_dispatch: bo
                         # buffer may correspond to multiple param buffers. If we zero out the entire
                         # grad buffer, it would clear the data of those param buffers that have not
                         # yet completed AG.
+                        if is_bf16_weight_group:
```
Comment (kunlunl) [inline megatron/core/distributed/distributed_data_parallel.py]: I think it's reasonable to skip `zero_` when the bucket is a bf16 bucket. But when copying from the main params to the "buffer", it doesn't consider bf16 or fp8, should bf16 also be skipped there (I mean the `_copy_main_params_to_param_buffer` function)?

File: megatron/core/distributed/param_and_grad_buffer.py [hunk 6]
```diff
@@ -343,9 +343,15 @@ def finish_param_sync(self, skip_next_bucket_dispatch: bool = False):
             # For the mxfp8_param with "reuse_grad_buf_for_mxfp8_param_ag=True",
             # we need to copy the param_data from the shared_param/grad_buffer to param.data
             # after the param all-gather.
+            is_bf16_weight_group = False
```
Comment (kunlunl) [inline megatron/core/distributed/param_and_grad_buffer.py]: Same question as above.

File: megatron/core/distributed/param_and_grad_buffer.py [hunk 7]
```diff
@@ -343,9 +343,15 @@ def finish_param_sync(self, skip_next_bucket_dispatch: bool = False):
...
             if self.ddp_config.reuse_grad_buf_for_mxfp8_param_ag:
                 for bucket in self.buckets:
                     for param in bucket.params:
+                        # Skip copying since bf16 weights in the mxfp8 model
+                        # are already mapped to param.data.
+                        if not is_mxfp8tensor(param) and not is_float8tensor(param):
```
Comment (kunlunl) [inline megatron/core/distributed/param_and_grad_buffer.py]: Same question as above.

File: megatron/core/distributed/param_and_grad_buffer.py [hunk 8]
```diff
@@ -820,8 +828,21 @@ def _does_param_require_new_bucket(param):
         cur_bucket_id = 0
         for param in params[::-1]:
             param_start_index, param_end_index, bucket_id = self.param_index_map[param]
-            # For MXFP8 param: we only need to map weight gradients to the buffer.
-            if not self.ddp_config.reuse_grad_buf_for_mxfp8_param_ag:
+            # For MXFP8 param:
+            # we only need to map bf16 weights (layernorm, embedding, etc) to the buffer.
+            if self.ddp_config.reuse_grad_buf_for_mxfp8_param_ag:
+                if not is_mxfp8tensor(param) and not is_float8tensor(param):
```
Comment (kunlunl) [inline megatron/core/distributed/param_and_grad_buffer.py]: Seems you can directly change the original condition from `if not self.ddp_config.reuse_grad_buf_for_mxfp8_param_ag` to `if not self.ddp_config.reuse_grad_buf_for_mxfp8_param_ag or not is_mxfp8tensor(param):`?

**Review by kunlunl — CHANGES_REQUESTED** [review]

File: megatron/core/distributed/distributed_data_parallel.py (hunk 2 above)
Comment (WanZzzzzz) [reply megatron/core/distributed/distributed_data_parallel.py]: Yes, removed it for simplicity.

File: megatron/core/distributed/distributed_data_parallel.py (hunk 4 above)
Comment (WanZzzzzz) [reply megatron/core/distributed/distributed_data_parallel.py]: Removed.

File: megatron/core/distributed/distributed_data_parallel.py (hunk 3 above)
Comment (WanZzzzzz) [reply megatron/core/distributed/distributed_data_parallel.py]: Yes, seems there are cases when a bucket group has both fp8 and bf16 buckets.
Made the judgement for each bucket.

File: megatron/core/distributed/distributed_data_parallel.py (hunk 5 above)
Comment (WanZzzzzz) [reply megatron/core/distributed/distributed_data_parallel.py]: No, we still need_copy_main_params_to_param_buffer function for bf16.
https://github.com/NVIDIA/Megatron-LM/blob/60a25aa67eae64a287d27a334eda36f37d9765ba/megatron/core/optimizer/optimizer.py#L584
When the reuse flag is true, we call  _copy_main_params_to_param_buffer instead of  _copy_main_params_to_model_params to cast master weights to bf16 buffer (which is mapped to bf16 params.)

File: megatron/core/distributed/param_and_grad_buffer.py (hunk 6 above)
Comment (WanZzzzzz) [reply megatron/core/distributed/param_and_grad_buffer.py]: Done

File: megatron/core/distributed/param_and_grad_buffer.py (hunk 7 above)
Comment (WanZzzzzz) [reply megatron/core/distributed/param_and_grad_buffer.py]: Done

File: megatron/core/distributed/param_and_grad_buffer.py (hunk 8 above)
Comment (WanZzzzzz) [reply megatron/core/distributed/param_and_grad_buffer.py]: Good point. Done

File: megatron/core/distributed/param_and_grad_buffer.py (hunk 1 above)
Comment (WanZzzzzz) [reply megatron/core/distributed/param_and_grad_buffer.py]: Avoided this according to @kunlunl 's suggestion above.

File: megatron/core/distributed/param_and_grad_buffer.py [hunk 9]
```diff
@@ -342,13 +342,21 @@ def finish_param_sync(self, skip_next_bucket_dispatch: bool = False):
 
             # For the mxfp8_param with "reuse_grad_buf_for_mxfp8_param_ag=True",
             # we need to copy the param_data from the shared_param/grad_buffer to param.data
-            # after the param all-gather.
+            # after the param all-gather.
```
Comment (kunlunl) [inline megatron/core/distributed/param_and_grad_buffer.py]: ~~Extra spaces.~~

**Comment by deepakn94** [comment]
Can we add a unit test for this?

**Comment by deepakn94** [comment]
/ok to test [6656cb2](https://github.com/NVIDIA/Megatron-LM/pull/3461/commits/6656cb2286a2f476ae564f440ec5a12d583ccfaf)

**Comment by deepakn94** [comment]
/ok to test [77a0371](https://github.com/NVIDIA/Megatron-LM/pull/3461/commits/77a03712f3f0200c803e998008a6bcd746416020)
